vowpalwabbit/config/options_cli.cc

```diff
@@ -0,0 +1,486 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "config/options_cli.h"
+
+#include <algorithm>
+#include <cctype>
+#include <iterator>
+#include <memory>
+#include <set>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <queue>
+
+#include "config/option.h"
+#include "config/options.h"
+
+#include "vw_exception.h"
+#include "vw_string_view.h"
+
+using namespace VW::config;
+
+bool is_number(const std::string& str)
+{
+  if (str.empty()) { return false; }
+
+  char* ptr;
+  std::strtof(str.c_str(), &ptr);
+  return (*ptr) == '\0';
+}
+
+bool is_number(VW::string_view str) { return is_number(std::string{str}); }
+
+template <typename T>
+std::vector<T> flatten_vectors(const std::vector<std::vector<T>>& vec_of_vecs)
+{
+  std::vector<T> result;
+  for (const auto& vec : vec_of_vecs) { result.insert(result.end(), vec.begin(), vec.end()); }
+  return result;
+}
+
+void check_disagreeing_option_values(const VW::string_view& ref_value, const std::string& option_name,
+    const std::vector<VW::string_view>& final_arguments)
+{
+  for (auto const& item : final_arguments)
+  {
+    if (item != ref_value)
+    {
+      std::stringstream ss;
+      ss << "Disagreeing option values for '" << option_name << "': '" << ref_value << "' vs '" << item << "'";
+      THROW_EX(VW::vw_argument_disagreement_exception, ss.str())
+    }
+  }
+}
+
+enum class option_type
+{
+  scalar,
+  boolean,
+  vector
+};
+
+option_type get_option_type(const base_option& option)
+{
+  if (option.m_type_hash == typeid(bool).hash_code()) { return option_type::boolean; }
+  if (option.m_type_hash == typeid(std::vector<std::string>).hash_code()) { return option_type::vector; }
+  return option_type::scalar;

```

Comment:
The problem with checking for scalar is it can be one of many types. typed_option has a static_assert which ensures scalar is the only other possible thing it can be at this point